### PR TITLE
Admin Flag for Disposal mobs

### PR DESCRIPTION
**Pull request by realforest2001**
# About the pull request
Allows admins to toggle a flag as to whether mobs will fit inside disposals tubes.

# Explain why it's good for the game
Regarding disposal tubes, there are some situations where I can see it being fitting to the situation, and this will also allow a measured level of return for a feature we had to remove due to abuse.
# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl:
add: Added an admin toggle flag for fitting mobs into disposals. (Within reason - no queens...)
add: The Crawler trait that allows mobs to enter vents will also allow them to climb into disposals regardless of above flag.
add: Added a subtype of disposal that can't fit mobs, used for the delivery units.
/:cl:

### CODE REVIEW

File: code/modules/recycling/disposal.dm [hunk 1]
```diff
@@ -128,23 +134,36 @@
...
+			/// Amazing what you can do with a bit of dexterity.
+			if(HAS_TRAIT(user, TRAIT_DEXTROUS))
+				max_grab_size++
+			/// Strong mobs can lift above their own weight.
+			if(HAS_TRAIT(user, TRAIT_SUPER_STRONG))//NB; this will mean Yautja can bodily lift MOB_SIZE_XENO(3) and Synths can lift MOB_SIZE_XENO_SMALL(2)
+				max_grab_size++
+			if(grabbed_mob.mob_size > max_grab_size || !(grabbed_mob.status_flags & CANPUSH))
+				to_chat(user, SPAN_WARNING("You don't have the strength to move [grabbed_mob]!"))
+				return FALSE//can't tighten your grip on mobs bigger than you and mobs you can't push.
+			else if(user.grab_level >= GRAB_AGGRESSIVE)
```
Comment (harryob) [inline code/modules/recycling/disposal.dm:154]: else if redundant after a return

File: code/modules/recycling/disposal.dm [hunk 2]
```diff
@@ -128,23 +134,36 @@
...
+				user.visible_message(SPAN_WARNING("[user] starts putting [grabbed_mob] into [src]."),
+				SPAN_WARNING("You start putting [grabbed_mob] into [src]."))
+				if(do_after(user, 20, INTERRUPT_ALL, BUSY_ICON_HOSTILE))
```
Comment (harryob) [inline code/modules/recycling/disposal.dm:157]: early return

File: code/modules/recycling/disposal.dm [hunk 3]
```diff
@@ -128,23 +134,36 @@
...
+					grabbed_mob.forceMove(src)
+					user.visible_message(SPAN_WARNING("[user] puts [grabbed_mob] into [src]."),
+					SPAN_WARNING("[user] puts [grabbed_mob] into [src]."))
+					user.attack_log += text("\[[time_stamp()]\] <font color='red'>Has placed [key_name(grabbed_mob)] in disposals.</font>")
+					grabbed_mob.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been placed in disposals by [user] ([user.ckey])</font>")
+					msg_admin_attack("[user] ([user.ckey]) placed [key_name(grabbed_mob)] in a disposals unit in [get_area(user)] ([user.loc.x],[user.loc.y],[user.loc.z]).", user.loc.x, user.loc.y, user.loc.z)
+					flush(TRUE)//Forcibly flushing someone if forced in by another player.
+			else
```
Comment (harryob) [inline code/modules/recycling/disposal.dm:165]: early return

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/recycling/disposal.dm [hunk 4]
```diff
@@ -37,6 +37,12 @@
 	active_power_usage = 3500
 	idle_power_usage = 100
 	var/disposal_pressure = 0
+	var/narrow_tube = FALSE
```
Comment (Zonespace27) [inline code/modules/recycling/disposal.dm:40]: codedoc

File: code/modules/recycling/disposal.dm [hunk 5]
```diff
@@ -128,23 +134,40 @@
...
+			/// Strong mobs can lift above their own weight.
+			if(HAS_TRAIT(user, TRAIT_SUPER_STRONG))//NB; this will mean Yautja can bodily lift MOB_SIZE_XENO(3) and Synths can lift MOB_SIZE_XENO_SMALL(2)
+				max_grab_size++
+			if(grabbed_mob.mob_size > max_grab_size || !(grabbed_mob.status_flags & CANPUSH))
+				to_chat(user, SPAN_WARNING("You don't have the strength to move [grabbed_mob]!"))
+				return FALSE//can't tighten your grip on mobs bigger than you and mobs you can't push.
+			if(user.grab_level >= GRAB_AGGRESSIVE)
+				user.visible_message(SPAN_WARNING("[user] starts putting [grabbed_mob] into [src]."),
+				SPAN_WARNING("You start putting [grabbed_mob] into [src]."))
+				if(!do_after(user, 20, INTERRUPT_ALL, BUSY_ICON_HOSTILE))
```
Comment (Zonespace27) [inline code/modules/recycling/disposal.dm:157]: seconds define

File: code/modules/recycling/disposal.dm [hunk 6]
```diff
@@ -161,51 +184,50 @@
...
-	if(!istype(target) || target.anchored || target.buckled || get_dist(user, src) > 1 || get_dist(user, target) > 1 || user.is_mob_incapacitated(TRUE) || isRemoteControlling(user) || target.mob_size >= MOB_SIZE_BIG)
-		return
-	if(!(ishuman(target)) || !(ishuman(user))) return
-	if(isanimal(user) && target != user) return //Animals cannot put mobs other than themselves into disposal
+	if((!MODE_HAS_TOGGLEABLE_FLAG(MODE_DISPOSABLE_MOBS) && !HAS_TRAIT(user, TRAIT_CRAWLER)) || narrow_tube)
+		to_chat(user, SPAN_WARNING("Looks a little bit too tight in there!"))
+		return FALSE
+
+	if(target != user)
+		to_chat(user, SPAN_WARNING("You need a better grip on [target] to force them into the [src]!"))
```
Comment (Zonespace27) [inline code/modules/recycling/disposal.dm:192]:
```suggestion
		to_chat(user, SPAN_WARNING("You need a better grip on [target] to force them into [src]!"))
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Review by harryob — APPROVED** [review]

**Pull request merged**